docs/guides/data-loading.md

```diff
@@ -161,6 +161,30 @@ export default function ProductCategory() {
 }
 ```
 
+If you are using TypeScript, you can use type inference to use Primsa Client generated types on when calling `useLoaderData`. This allowes better type safety and intellisense when writing your code that uses the loaded data.
+
+```tsx filename=tsx filename=app/routes/products/$productId.tsx
+import { useLoaderData } from "remix";
+import { db } from "~/db.server";
+
+type LoaderData = Awaited<ReturnType<typeof loader>>
+
+export let loader = async ({ params }) => {
+  return db.product.findMany({});
+};
```

Comment:
The challenge here is that if you want to set headers you can no longer use the return value as it's a Response object instead. So how about this:

```suggestion
import { useLoaderData, json } from "remix";
import { db } from "~/db.server";

type LoaderData = Awaited<ReturnType<typeof getLoaderData>>

async function getLoaderData() {
  const products = await db.product.findMany({
    select: {
      id: true,
      name: true,
      imgSrc: true,
    }
  })
  return { products }
}

export let loader = async () => {
  return json<LoaderData>(await getLoaderData());
};
```